### ISSUE TITLE
onSuggestionSelected not called on blur

### ISSUE THREAD

**Issue by ZebraFlesh** [moroshko/react-autosuggest#213]
I'm providing a function for the `onSuggestionSelected` prop. However, this isn't called when the input field blurs. This leads to some weird state management because `inputProps.onChange` was called -- it looks like the underlying value was selected, but because my callback wasn't invoked that's not actually the case.

**Comment by moroshko** [comment]
Could you create a Codepen that demonstrates the issue?

**Comment by ZebraFlesh** [comment]
Sure! http://codepen.io/anon/pen/wWYAJY?editors=0010

The problem is that I can't work around this issue with the publicly available API. I could add an `onBlur` handler to the `inputProps`, but that function doesn't have access to the correct information to properly invoke `onSuggestionSelected` (basically, everything in the second parameter of that function).

**Comment by moroshko** [comment]
Thanks for the Codepen, it's an interesting one!

I tried to see what Google does in this situation, and apparently they don't activate the search in this case.

Can I ask what are you doing in your app when <kbd>Tab</kbd> is pressed?

**Comment by ZebraFlesh** [comment]
We're using react-autosuggest to display autocomplete results from Google Places. Each suggestion is a "thin" data object returned by the Google Places API and contains just some display text and a lookup ID. Selecting a suggestion causes a followup call to the Google Places API using that lookup ID in order to obtain the full data object about a given suggestion. Due to how this API is designed, we'll never have the necessary data until the user makes a selection and we perform that second ajax request.

When the user blurs the field, we're forced to compare the value of the input against the list of suggestions to determine if a suggestion was highlighted and then manually invoke `onSuggestionSelected`. Looking at react-autosuggest, it already has all of the necessary data for this scenario -- it's just not in the public API.

I think any of the following changes would help solve this problem:
1. invoke `onSuggestionSelected` on blur
2. provide an option to invoke `onSuggestionSelected` on blur
3. provide the focused suggestion object as a parameter to the blur event handler
4. provide a public read-only version of the `focusedSectionIndex` and `focusedSuggestionIndex` props

Of those, (2) might be best the compromise: it would be opt-in functionality, so it wouldn't break any existing implementations.

**Comment by moroshko** [comment]
Thanks for the detailed explanation!
Semantically, when users blur the input field, they are not necessarily selecting a suggestion. Your scenario of using the <kbd>Up</kbd>/<kbd>Down</kbd> keys and then pressing <kbd>Tab</kbd> is probably the closest one that can be considered as an intentional suggestion selection. But, if user just clicks on some other element on the page, they are not necessarily expressing the intent of selecting a suggestion. Therefore, I prefer (3) over (1) and (2).

What do you think about something like this?

```js
inputProps.onBlur = (event, { focusedSuggestion }) => {
  if (focusedSuggestion !== null) {
    // call Google Places API with focusedSuggestion.lookupID
  }
}
```

**Comment by ZebraFlesh** [comment]
I think that solution would certainly get the job done. It's a good compromise that offers a hook for those that want to implement select on blur semantics, but it doesn't force it on everyone.

**Comment by moroshko** [comment]
Would you want to call the Google Places API if user focused a suggestion using <kbd>Up</kbd>/<kbd>Down</kbd>, and then clicked on some other element on the page (rather than pressing <kbd>Tab</kbd>)?

I'm thinking maybe providing a `method` (like we do in [`inputProps.onChange`](https://github.com/moroshko/react-autosuggest#inputPropsProp)) will be useful to differentiate between various scenarios. So, if you wanted to call the Google Places API only when the input is blurred using <kbd>Tab</kbd>, you could do:

```js
inputProps.onBlur = (event, { focusedSuggestion, method }) => {
  if (method === 'tab' && focusedSuggestion !== null) {
    // call Google Places API with focusedSuggestion.lookupID
  }
}
```

What do you think?

**Comment by ZebraFlesh** [comment]
That part goes beyond my use case: regardless of how the field is blurred, I'm going to take the focused element as a selection. I think that's also web semantics: blur is blur, no matter how it was invoked. (Maybe that's a naive statement?)

**Comment by moroshko** [comment]
OK, let's add `focusedSuggestion` for now, and we can always add `method` later if there is a valid use case for it. I hope to free up some time in the next couple of days to implement this.

**Comment by moroshko** [comment]
`focusedSuggestion` is now available in [v5.1.0](https://github.com/moroshko/react-autosuggest/releases/tag/v5.1.0).
